File: typescript/tests/capability_registry.test.ts
// capability_registry.test.ts
//
// Verifies the ported ExternalCapabilityRegistry (CapabilityRegistry.cs): the
// full 32-entry set, case-insensitive Find, and ByPackage filtering.

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  EXTERNAL_CAPABILITY_REGISTRY,
  ExternalCapabilityRegistry,
  findCapability,
  capabilitiesByPackage,
} from '../src/companion/capability_registry';

describe('ExternalCapabilityRegistry', () => {
  it('contains all 30 earmarked capabilities', () => {
    assert.equal(EXTERNAL_CAPABILITY_REGISTRY.length, 30);
    assert.equal(ExternalCapabilityRegistry.all.length, 30);
  });

  it('every entry has an id, license, strategy, target package, and >=1 bullet', () => {
    for (const c of EXTERNAL_CAPABILITY_REGISTRY) {
      assert.ok(c.id.length > 0, `id for ${c.id}`);
      assert.ok(c.license.length > 0, `license for ${c.id}`);
      assert.ok(['vendor', 'pattern-port', 'wrap'].includes(c.strategy), `strategy for ${c.id}`);
      assert.ok(c.targetPackage.startsWith('CircleAI.'), `package for ${c.id}`);
      assert.ok(c.valueBullets.length >= 1, `bullets for ${c.id}`);
    }
  });

  it('finds by id case-insensitively', () => {
    const a = findCapability('claude-mem');
    assert.ok(a);
    assert.equal(a.repo, 'thedotmack/claude-mem');
    assert.equal(a.license, 'MIT');
    assert.equal(a.strategy, 'pattern-port');
    assert.equal(a.targetPackage, 'CircleAI.Memory');
    // Case-insensitive.
    assert.equal(findCapability('AMPHION')?.id, 'Amphion');
    assert.equal(ExternalCapabilityRegistry.find('hippoRAG')?.id, 'HippoRAG');
  });

  it('returns null for an unknown id', () => {
    assert.equal(findCapability('does-not-exist'), null);
  });

  it('preserves a known multi-bullet entry verbatim', () => {
    const superpowers = findCapability('superpowers');
    assert.ok(superpowers);
    assert.equal(superpowers.valueBullets.length, 8);
    assert.equal(superpowers.valueBullets[6], 'TDD RED-GREEN-REFACTOR mandatory gate');
  });

  it('lists entries by target package (case-insensitive)', () => {
    // Two Games entries: aimangastudio + flame.
    const games = capabilitiesByPackage('CircleAI.Games');
    assert.deepEqual(
      games.map((c) => c.id).sort(),
      ['aimangastudio', 'flame'],
    );
    // Two Speech entries: Amphion + yapsnap.
    assert.equal(ExternalCapabilityRegistry.byPackage('circleai.speech').length, 2);
    // Two Inference entries: airllm + shard.
    assert.equal(capabilitiesByPackage('CircleAI.Inference').length, 2);
  });

  it('unknown package → empty list', () => {
    assert.equal(capabilitiesByPackage('CircleAI.Nope').length, 0);
  });
});

File: rust/tests/capability_registry_test.rs
//! capability_registry_test.rs
//!
//! Verifies ExternalCapabilityRegistry: the full absorption catalogue, id lookup,
//! and by-package filtering. Mirrors the C# ExternalCapabilityRegistry.

use circle_ai::companion::capability_registry::ExternalCapabilityRegistry;

#[test]
fn all_has_the_full_catalogue() {
    let all = ExternalCapabilityRegistry::all();
    // The C# registry has 30 entries.
    assert_eq!(all.len(), 30);
    // Every entry has a non-empty id, license, strategy, target package, and at
    // least one value bullet.
    for c in &all {
        assert!(!c.id.is_empty());
        assert!(!c.license.is_empty());
        assert!(!c.strategy.is_empty());
        assert!(!c.target_package.is_empty());
        assert!(!c.value_bullets.is_empty());
    }
}

#[test]
fn find_is_case_insensitive() {
    let hit = ExternalCapabilityRegistry::find("HippoRAG").unwrap();
    assert_eq!(hit.id, "HippoRAG");
    assert_eq!(hit.repo.as_deref(), Some("OSU-NLP-Group/HippoRAG"));
    assert_eq!(hit.target_package, "CircleAI.Memory.HippoRAG");
    // Different casing still matches.
    assert!(ExternalCapabilityRegistry::find("hipporag").is_some());
    assert!(ExternalCapabilityRegistry::find("no-such-cap").is_none());
}

#[test]
fn by_package_filters_correctly() {
    // Two capabilities target CircleAI.Speech (Amphion + yapsnap).
    let speech = ExternalCapabilityRegistry::by_package("CircleAI.Speech");
    assert_eq!(speech.len(), 2);
    let ids: Vec<&str> = speech.iter().map(|c| c.id.as_str()).collect();
    assert!(ids.contains(&"Amphion"));
    assert!(ids.contains(&"yapsnap"));
    // Two target CircleAI.Inference (airllm + shard).
    let inference = ExternalCapabilityRegistry::by_package("CircleAI.Inference");
    assert_eq!(inference.len(), 2);
    // Case-insensitive package match.
    assert_eq!(
        ExternalCapabilityRegistry::by_package("circleai.speech").len(),
        2
    );
    // Unknown package → empty.
    assert!(ExternalCapabilityRegistry::by_package("CircleAI.Nope").is_empty());
}

#[test]
fn claude_mem_entry_is_intact() {
    let c = ExternalCapabilityRegistry::find("claude-mem").unwrap();
    assert_eq!(c.license, "MIT");
    assert_eq!(c.strategy, "pattern-port");
    assert_eq!(c.target_package, "CircleAI.Memory");
    assert_eq!(c.value_bullets.len(), 10);
    assert_eq!(c.value_bullets[0], "Multi-platform memory adapter");
}
